Source/WebCore/style/RuleFeature.cpp

```diff
@@ -371,10 +372,24 @@ static PseudoClassInvalidationKey makePseudoClassInvalidationKey(CSSSelector::Ps
     return makePseudoClassInvalidationKey(pseudoClassType, InvalidationKeyType::Universal);
 };
 
-void RuleFeatureSet::collectFeatures(const RuleData& ruleData)
+void RuleFeatureSet::collectFeatures(const RuleData& ruleData, const Vector<Ref<const StyleRuleScope>>& scopeRules)
 {
     SelectorFeatures selectorFeatures;
     recursivelyCollectFeaturesFromSelector(selectorFeatures, *ruleData.selector());
+
+    for (auto& scopeRule : scopeRules) {
+        auto collectSelectorList = [&] (const auto& selectorList) {
+            if (!selectorList.isEmpty()) {
+                for (const auto* subSelector = selectorList.first() ; subSelector; subSelector = CSSSelectorList::next(subSelector)) {
+                    recursivelyCollectFeaturesFromSelector(selectorFeatures, *subSelector, MatchElement::Ancestor);
+                    recursivelyCollectFeaturesFromSelector(selectorFeatures, *subSelector, MatchElement::Subject);
```

Comment:
You could add MatchElement::AncestorOrSubject to avoid adding the selector twice. I think the implementation should be largely self-explanatory. Or more specific MatchElement::ScopeStart/ScopeEnd.

Doesn't need to be in this patch.